dallinger/mturk.py

```diff
@@ -0,0 +1,171 @@
+import datetime
+from boto.mturk.connection import MTurkConnection
+from boto.mturk.price import Price
+from boto.mturk.qualification import LocaleRequirement
+from boto.mturk.qualification import PercentAssignmentsApprovedRequirement
+from boto.mturk.qualification import Qualifications
+from boto.mturk.question import ExternalQuestion
+
+
+def timestr_to_dt(timestr):
+    return datetime.datetime.strptime(timestr, '%Y-%m-%dT%H:%M:%SZ')
+
+
+class MTurkServiceException(Exception):
+    """Custom exception type"""
+
+
+class MTurkService(object):
+    """Facade for Amazon Mechanical Turk services provided via the boto
+       library.
+    """
+    production_mturk_server = 'mechanicalturk.amazonaws.com'
+    sandbox_mturk_server = 'mechanicalturk.sandbox.amazonaws.com'
+    _connection = None
+
+    def __init__(self, aws_access_key_id, aws_secret_access_key, sandbox=True):
+        self.aws_access_key_id = aws_access_key_id
+        self.aws_secret_access_key = aws_secret_access_key
+        self.is_sandbox = sandbox
+
+    @property
+    def mturk(self):
+        """Cached MTurkConnection"""
+        if not self.aws_access_key_id or not self.aws_secret_access_key:
+            raise MTurkServiceException('AWS access key and secret not set.')
+        login_params = {
+            'aws_access_key_id': self.aws_access_key_id,
+            'aws_secret_access_key': self.aws_secret_access_key,
+            'host': self.host
+        }
+        if self._connection is None:
+            self._connection = MTurkConnection(**login_params)
+        return self._connection
```

Comment:
@jessesnyder This caching would be cleaner if we borrow Pyramid's `reify` decorator: http://docs.pylonsproject.org/projects/pyramid/en/latest/_modules/pyramid/decorator.html#reify